executor/insert_common.go

```diff
@@ -368,102 +352,115 @@ func (e *InsertValues) getColDefaultValue(idx int, col *table.Column) (d types.D
 	return defaultVal, nil
 }
 
-// initDefaultValues fills generated columns, auto_increment column and empty column.
+// fillColValue fills the column value if it is not set in the insert statement.
+func (e *InsertValues) fillColValue(datum types.Datum, idx int, column *table.Column, hasValue bool) (types.Datum,
+	error) {
+	if mysql.HasAutoIncrementFlag(column.Flag) {
+		d, err := e.adjustAutoIncrementDatum(datum, hasValue, column)
+		if err != nil {
+			return types.Datum{}, errors.Trace(err)
+		}
+		return d, nil
+	} else if !hasValue {
+		d, err := e.getColDefaultValue(idx, column)
+		if e.filterErr(err) != nil {
+			return types.Datum{}, errors.Trace(err)
+		}
+		return d, nil
+	}
+	return datum, nil
+}
+
+// fillRow fills generated columns, auto_increment column and empty column.
 // For NOT NULL column, it will return error or use zero value based on sql_mode.
-func (e *InsertValues) initDefaultValues(row []types.Datum, hasValue []bool) error {
+func (e *InsertValues) fillRow(cols []*table.Column, row []types.Datum, hasValue []bool, valLen int) ([]types.Datum, error) {
 	for i, c := range e.Table.Cols() {
```

Comment:
There are `cols` from caller and `e.Table.Cols()`, which one should we use?

Reply:
Good catch! I think it may contain some bugs.